include/crocoddyl/core/integrator/rk4.hxx

```diff
@@ -0,0 +1,257 @@
+                                               const Eigen::Ref<const VectorXs>& u) {
+  if (static_cast<std::size_t>(x.size()) != state_->get_nx()) {
+    throw_pretty("Invalid argument: "
+                 << "x has wrong dimension (it should be " + std::to_string(state_->get_nx()) + ")");
+  }
+  if (static_cast<std::size_t>(u.size()) != nu_) {
+    throw_pretty("Invalid argument: "
+                 << "u has wrong dimension (it should be " + std::to_string(nu_) + ")");
+  }
+
+  const std::size_t& nv = differential_->get_state()->get_nv();
+
+  // Static casting the data
+  boost::shared_ptr<Data> d = boost::static_pointer_cast<Data>(data);
+
+  // Computing the acceleration and cost
+  differential_->calc(d->differential[0], x, u);
+
+  // Computing the next state (discrete time)
+  if (enable_integration_) {
+    d->y[0] = x;
+    d->ki[0].head(nv) = d->y[0].tail(nv);
+    d->ki[0].tail(nv) = d->differential[0]->xout;
+    d->integral[0] = d->differential[0]->cost;
+    for (std::size_t i = 1; i < 4; ++i) {
+      d->dx_rk4[i].noalias() = time_step_ * rk4_c_[i] * d->ki[i - 1];
+      differential_->get_state()->integrate(x, d->dx_rk4[i], d->y[i]);
+      differential_->calc(d->differential[i], d->y[i], u);
+      d->ki[i].head(nv) = d->y[i].tail(nv);
+      d->ki[i].tail(nv) = d->differential[i]->xout;
+      d->integral[i] = d->differential[i]->cost;
+    }
+    d->dx = (d->ki[0] + Scalar(2.) * d->ki[1] + Scalar(2.) * d->ki[2] + d->ki[3]) * time_step_ / Scalar(6.);
+    differential_->get_state()->integrate(x, d->dx, d->xnext);
+    d->cost = (d->integral[0] + Scalar(2.) * d->integral[1] + Scalar(2.) * d->integral[2] + d->integral[3]) *
+              time_step_ / Scalar(6.);
+  } else {
+    d->dx.setZero();
+    d->xnext = x;
+    d->cost = d->differential[0]->cost;
+  }
+
+  // Updating the cost value
+  if (with_cost_residual_) {
+    d->r = d->differential[0]->r;
+  }
+}
+
+template <typename Scalar>
+void IntegratedActionModelRK4Tpl<Scalar>::calcDiff(const boost::shared_ptr<ActionDataAbstract>& data,
+                                                   const Eigen::Ref<const VectorXs>& x,
+                                                   const Eigen::Ref<const VectorXs>& u) {
+  if (static_cast<std::size_t>(x.size()) != state_->get_nx()) {
+    throw_pretty("Invalid argument: "
+                 << "x has wrong dimension (it should be " + std::to_string(state_->get_nx()) + ")");
+  }
+  if (static_cast<std::size_t>(u.size()) != nu_) {
+    throw_pretty("Invalid argument: "
+                 << "u has wrong dimension (it should be " + std::to_string(nu_) + ")");
+  }
+
+  const std::size_t& nv = differential_->get_state()->get_nv();
+
+  boost::shared_ptr<Data> d = boost::static_pointer_cast<Data>(data);
+
+  differential_->calcDiff(d->differential[0], x, u);
+
+  if (enable_integration_) {
+    d->dki_dy[0].topRightCorner(nv, nv).diagonal().array() = (Scalar)1;
```

Comment:
I will move this operation to the data creation, so we avoid to fill the same value each time

Reply:
done